### ISSUE TITLE
Can't pm-update with Drush 8.3.3 | PHP 7.2 and PHP 7.4 installed

### ISSUE THREAD

**Issue by mataski** [drush-ops/drush#4414]
**Describe the bug**
Can't run pm-update `pm-updatestatus failed.`

**To Reproduce**
Update Drush to latest version 8.3.3 together with Composer version 1.10.6 and run `drush pm-update`. It stopped working after this shared hosting provider installed PHP 7.4 while website still runs on 7.2, but `drush <any_command>` picks up 7.4 by default and gives error:

```
"continue" targeting switch is equivalent to "break". Did you mean to use "continue 2"? userprotect.module:627                                                                           [warning]
"continue 2" targeting switch is equivalent to "break 2". Did you mean to use "continue 3"? userprotect.module:647                                                                       [warning]
Deprecated function: Array and string offset access syntax with curly braces is deprecated in include_once() (line 20 of /users/asd/htdocs/includes/file.phar.inc).           [error]
```

So you have to run it through older version of PHP and if you do this `whereis php` you get:

`php: /usr/bin/php7.0 /usr/bin/php7.3 /usr/bin/php /usr/bin/php7.2 /usr/bin/php5.6 /usr/bin/php7.4 /usr/lib/php /etc/php /usr/include/php /usr/share/php7.3-json /usr/share/php5.6-sybase /usr/share/php7.4-pspell /usr/share/php7.0-mbstring /usr/share/php7.2-gd /usr/share/php7.3-soap /usr/share/php7.4-sqlite3 /usr/share/php7.4-soap /usr/share/php7.3-mysql /usr/share/php7.0-sqlite3 /usr/share/php7.0-soap /usr/share/php5.6-bcmath /usr/share/php7.1-xml /usr/share/php7.2-sybase /usr/share/php7.0-imap /usr/share/php7.4-gd /usr/share/php5.6-xmlrpc /usr/share/php7.2-soap /usr/share/php7.4-mbstring /usr/share/php7.0-readline /usr/share/php5.6-zip /usr/share/php5.6-intl /usr/share/php7.1-common /usr/share/php7.3-gd /usr/share/php7.4-bcmath /usr/share/php5.6-soap /usr/share/php7.3-common /usr/share/php5.6-mcrypt /usr/share/php7.0-sybase /usr/share/php7.2-xmlrpc /usr/share/php7.4-xml /usr/share/php7.2-json /usr/share/php7.0-mysql /usr/share/php5.6-gd /usr/share/php7.2-imap /usr/share/php7.4-common /usr/share/php7.2-mbstring /usr/share/php7.2-common /usr/share/php /usr/share/php7.3-xmlrpc /usr/share/php7.2-pspell /usr/share/php7.3-zip /usr/share/php7.0-intl /usr/share/php7.3-mbstring /usr/share/php7.3-readline /usr/share/php7.2-opcache /usr/share/php5.6-opcache /usr/share/php7.3-opcache /usr/share/php7.0-common /usr/share/php5.6-common /usr/share/php7.4-imap /usr/share/php7.0-gd /usr/share/php7.2-mysql /usr/share/php7.4-mysql /usr/share/php5.6-sqlite3 /usr/share/php7.0-mcrypt /usr/share/php7.2-intl /usr/share/php7.3-xml /usr/share/php7.4-curl /usr/share/php7.0-zip /usr/share/php5.6-mysql /usr/share/php7.4-opcache /usr/share/php5.6-mbstring /usr/share/php7.3-bcmath /usr/share/php7.4-json /usr/share/php7.4-xmlrpc /usr/share/php7.2-bcmath /usr/share/php7.4-intl /usr/share/php7.2-xml /usr/share/php7.0-pspell /usr/share/php7.0-opcache /usr/share/php7.4-sybase /usr/share/php7.2-sqlite3 /usr/share/php5.6-curl /usr/share/php7.0-json /usr/share/php5.6-pspell /usr/share/php7.2-zip /usr/share/php7.3-pspell /usr/share/php7.0-xmlrpc /usr/share/php7.2-curl /usr/share/php7.3-sybase /usr/share/php5.6-imap /usr/share/php7.3-intl /usr/share/php5.6-json /usr/share/php7.4-zip /usr/share/php7.0-xml /usr/share/php7.0-bcmath /usr/share/php7.3-imap /usr/share/php7.3-sqlite3 /usr/share/php7.2-readline /usr/share/php7.0-curl /usr/share/php7.3-curl /usr/share/php5.6-xml /usr/share/php5.6-readline /usr/share/php7.4-readline /usr/share/man/man1/php.1.gz`

Which means the old working version of 7.2 is at ` /usr/bin/php7.2` so you run something like `/usr/bin/php7.2 ~/composer/vendor/drush/drush/drush.php status` and it works:

```
 Drupal version                  :  7.67
 Site URI                        :  http://default
 Database driver                 :  mysql
 Database hostname               :  localhost
 Database port                   :
 Database username               :  asd
 Database name                   :  asd
 Database                        :  Connected
 Drupal bootstrap                :  Successful
 Drupal user                     :
 Default theme                   :  asd
 Administration theme            :  rubik
 PHP configuration               :  /etc/php/7.2/cli/php.ini
 PHP OS                          :  Linux
 Drush script                    :  /users/asd/composer/vendor/drush/drush/drush.php
 Drush version                   :  8.3.3
 Drush temp directory            :  /tmp
 Drush configuration             :
 Drush alias files               :
 Install profile                 :  standard
 Drupal root                     :  /users/asd/htdocs
 Drupal Settings File            :  sites/default/settings.php
 Site path                       :  sites/default
 File directory path             :  sites/default/files
 Private file directory path     :  ../private
 Temporary file directory path   :  ../tmp
```

But if you run `/usr/bin/php7.2 ~/composer/vendor/drush/drush/drush.php pm-update` you get `pm-updatestatus failed.`

**Expected behavior**
Expect to update modules and core.

**Actual behavior**
It failed.

**Workaround**
Not aware of it.

### System Configuration
| Q               | A
| --------------- | ---
| Drush version?  | 8.3.3
| Drupal version? | 7.67
| PHP version     | 7.4 and below
| OS?             | Linux

**Additional information**
I had this problem with old Drush version 8.1.15 and Composer version 1.5.5 before updates. Everything worked fine until shared hosting pushed PHP 7.4 and now it's bricked, like at least 20 websites can't update.

Thank you for great tool.

**Comment by greg-1-anderson** [comment]
When you install something with Composer, the solver is going to look for dependencies that are compatible with the current version of PHP that you are running. This means that you might get different dependencies if you install Drush with PHP 7.2 vs if you install it with PHP 7.4. So, if you install Drush with PHP 7.2 and run it on PHP 7.4, you're not necessarily going to get correct results.

So the interesting question that this issue brings up is, what should we do about the drush.phar? We have to pick some version of PHP to use to generate the dependencies that are bundled in the .phar, and yet, Drush currently claims to support PHP 5.4 through 7.4.  This is getting to be increasingly difficult.

Also, while we are on this topic, Drush claims to support PHP 7.4, but there are no tests running on that version. Perhaps we should start there, and see if the 7.4 tests even pass. It could be that this situation is simply due to an oversight that we can resolve.

**Comment by greg-1-anderson** [comment]
#4415 worked pretty well; there were no signs of the symptoms reported in the OP.

Could you submit a failing test that demonstrates the problem?

**Comment by mataski** [comment]
What is a failing test and how do you submit that?

**Comment by mataski** [comment]
Anyway here's a screencap trying to reproduce on same machine but under different users and different websites, same structure, same everything with same versions, it works on one site, but not the other and so on.

https://streamable.com/oto3n8

**Comment by greg-1-anderson** [comment]
Here is an example of a passing set of tests running on PHP 7.4 with Drush 8:

https://travis-ci.org/github/drush-ops/drush/jobs/689259602

To make a failing test, just submit a regular pull request that modifies the existing tests. There are already tests that exercise updatedb and pass on PHP 7.4.  Copy one of these, and change the operations it does to match what you're doing in your screencap. If the test fails on Travis, then you have successfully submitted a failing test that reproduces your problem.

**Comment by greg-1-anderson** [comment]
Oh, you are still on Drupal 7.  Drupal 7 does not support PHP 7.4 yet.  Upgrade to 7.70, and then search for the PHP 7.4 issues.  Apply those to your site, and see if updatedb still fails.

**Comment by greg-1-anderson** [comment]
It looks like HEAD of the Drupal 7 branch now works with PHP 7.4:

https://www.drupal.org/project/drupal/issues/3081386

So, you could test this by using the latest dev release of Drupal 7. For production, it would probably be better to use an earlier version of PHP and a stable release of Drupal.

**Comment by mataski** [comment]
I did upgrade to 7.70 manually and still get same issue, `pm-update` won't run, both 7.2 and 7.4 PHP versions.

7.2 gives `pm-updatestatus failed.`

7.4 gives
```
"continue" targeting switch is equivalent to "break". Did you mean to use "continue 2"? userprotect.module:627                                                                           [warning]
"continue 2" targeting switch is equivalent to "break 2". Did you mean to use "continue 3"? userprotect.module:647                                                                       [warning]
Deprecated function: Array and string offset access syntax with curly braces is deprecated in include_once() (line 20 of /users/asd/htdocs/includes/file.phar.inc).           [error]
```

Drush 8.3.3

**Comment by mataski** [comment]
This patch https://www.drupal.org/project/drupal/issues/3105433 fix the issue, I don't know why.

It is annoying tho, you have to apply it first if you want to update to 7.70 with `drush pm-update` and after the update it gets overwritten so you have to reapply same patch again if you want `drush pm-update` to work with 7.70.

I just don't understand why I'm the only one getting this problem on dozens of websites.

**Comment by greg-1-anderson** [comment]
The patch you reference should not have any effect on php 7.2 behavior.

Have you tried with Drush 8.3.2, and checked to see if the behavior is any different? On PHP 7.2, that is, since 7.4 support was added in Drush 8.3.3.

**Comment by mataski** [comment]
I think that patch has effect on 7.4. Like I wrote before Drush didn't work on both 7.2 and 7.4 after the hosting implemented the 7.4. And it didn't matter 8.3.3 or lower version.

Anyway, so I just dropped 7.2 and moved with 7.4 now.

You have to apply the patch twice.
1. apply the patch.
2. `drush pm-update`
3. update only downloads the files and fails the `updatedb` because the patch was already overwritten by new update e.g. 7.70
4. reapply the patch.
5. `drush updatedb` works.

So 7.70 doesn't solve anything, you still have to apply the patch before and after.

I think only --dev version of Drupal 7 works without patching considering it was committed.
